csrc/optimization/alias_analysis.cpp

```diff
@@ -33,10 +33,58 @@ bool isContiguous(const TensorView& tv) {
   return true;
 }
 
-// Finds aliases of `root` and stores the findings in `alias_to_source`.
-void findAliasesOfRoot(
-    const TensorView* root,
-    AliasAnalysisResult& alias_to_source) {
+// Whether a ViewOp transforms any expanded broadcast IterDomain in the input.
+// This is a corner case in which we can't always turn `out` into an alias.
+//
+// For example, given
+//
+//   t0 = makeContigConcreteTensor({4, 5});
+//   t1 = broadcast(t0, {false, false, true});
+//   t2 = expand(t1, {4, 5, 6});
+//
+// `reshape(t2, {40, 3})` and `reshape(t2, {4, 30})` have to copy data.
+// `reshape(t2, {40, 3})` splits the expanded broadcast IterDomain (which is 6)
+// into 2 and 3 and merges the 2 with preceding IterDomains. `reshape(t2, {4,
+// 30})` merges the expanded broadcast IterDomain. However, the output of
+// `reshape(t2, {20, 6})` can simply be an alias because the expanded broadcast
+// IterDomain is forwarded not transformed.
+//
+// As a future improvement, when an expanded broadcast dimension is only split,
+// the output of the reshape can be an alias. However, nvFuser currently decides
+// to materialize the expansion, making the output not an alias (#1126).
+//
+// Obviously, this function assumes `in` and `out` are the input and output
+// TensorView of the same ViewOp.
+bool transformsExpandedBroadcastIterDomain(TensorView* in, TensorView* out) {
```

Comment:
This feels like a general analysis that could be added as a `TensorDomain` method.

Let `t` and `f` denote contiguous and non-contiguous `Iteration` domains and let `b` and `e`denote regular broadcast expanded broadcast to some non-unit size. The output IDs follow these rules I believe (please double check), where `C` denotes a required copy because the output cannot be expressed with a single stride:
```
Split:
  t -> (t, t)
  f -> (t, f)
  b -> (b, b)  <- split factor must be 1
  e -> (e, e), (b, e), or (e, b) depending on split factor (after #1126) -- last two are BroadcastOp

Merge:
  o i ->
   o\i | t  f  b  e
   ----+-----------
    t  | t  f  t  C
    f  | C  C  f  C
    b  | t  f  b  e
    e  | C  C  e  e
```
Statuses `b` and `e` are represented in the IR with `isBroadcast()` and `hasExpandedExtent()`. You'd have to propagate contiguity `t` and `f` since that is not currently tracked at the `IterDomain` level but instead it is a `TensorDomain` property. When you encounter a `C`, you can stop propagating since you know you can't alias at that point.

Note that this is not quite symmetric: `Merge t f` can be no-copy since the output just uses the inner stride, but `Merge f t` consists of disjointed contiguous segments so it can't be handled with one stride parameter.

It's possible that transforms generating `C` outputs could be later inverted, but `analyzeView` should not create such sequences.

Reply:
@jacobhinkle Not exactly sure what you mean by:

> b -> (b, b) <- split factor must be 1

We don't enforce any restriction of splitting broadcast domains. For example, it's possible to split by, say, 4, which would create a broadcast domain of size 4. It's not marked as expanded IIRC.

Reply:
> We don't enforce any restriction of splitting broadcast domains.

We don't enforce it on the scheduling op `split()`, but in a reshape wouldn't we always be preserving numel in each  `SplitTransform` and `MergeTransform` transform?

Reply:
Yes, broadcast domains should be squeezed before reshape transformations, except for expanded domains